locking/locks.go

```diff
@@ -0,0 +1,198 @@
+	// various command implementations.
+	API = api.NewClient(nil)
+	// errNoMatchingLocks is an error returned when no matching locks were
+	// able to be resolved
+	errNoMatchingLocks = errors.New("lfs: no matching locks found")
+	// errLockAmbiguous is an error returned when multiple matching locks
+	// were found
+	errLockAmbiguous = errors.New("lfs: multiple locks found; ambiguous")
+)
+
+// Lock attempts to lock a file on the given remote name
+// path must be relative to the root of the repository
+// Returns the lock id if successful, or an error
+func Lock(path, remote string) (id string, e error) {
+	// TODO: API currently relies on config.Config but should really pass to client in future
+	savedRemote := config.Config.CurrentRemote
+	config.Config.CurrentRemote = remote
+	defer func() { config.Config.CurrentRemote = savedRemote }()
+
+	// TODO: this is not really the constraint we need to avoid merges, improve as per proposal
+	latest, err := git.CurrentRemoteRef()
+	if err != nil {
+		return "", err
+	}
+
+	s, resp := API.Locks.Lock(&api.LockRequest{
+		Path:               path,
+		Committer:          api.CurrentCommitter(),
+		LatestRemoteCommit: latest.Sha,
+	})
+
+	if _, err := API.Do(s); err != nil {
+		return "", fmt.Errorf("Error communicating with LFS API: %v", err)
+	}
+
+	if len(resp.Err) > 0 {
+		return "", fmt.Errorf("Server unable to create lock: %v", resp.Err)
+	}
+
+	if err := cacheLock(resp.Lock.Path, resp.Lock.Id); err != nil {
+		return "", fmt.Errorf("Error caching lock information: %v", err)
+	}
+
+	return resp.Lock.Id, nil
+}
+
+// Unlock attempts to unlock a file on the given remote name
+// path must be relative to the root of the repository
+// Force causes the file to be unlocked from other users as well
+func Unlock(path, remote string, force bool) error {
+	// TODO: API currently relies on config.Config but should really pass to client in future
+	savedRemote := config.Config.CurrentRemote
+	config.Config.CurrentRemote = remote
+	defer func() { config.Config.CurrentRemote = savedRemote }()
+
+	id, err := lockIdFromPath(path)
+	if err != nil {
+		return fmt.Errorf("Unable to get lock id: %v", err)
+	}
+
+	return UnlockById(id, remote, force)
+
+}
+
+// Unlock attempts to unlock a lock with a given id on the remote
+// Force causes the file to be unlocked from other users as well
+func UnlockById(id, remote string, force bool) error {
+	s, resp := API.Locks.Unlock(id, force)
+
+	if _, err := API.Do(s); err != nil {
+		return fmt.Errorf("Error communicating with LFS API: %v", err)
+	}
+
+	if len(resp.Err) > 0 {
+		return fmt.Errorf("Server unable to unlock lock: %v", resp.Err)
+	}
+
+	if err := cacheUnlockById(id); err != nil {
+		return fmt.Errorf("Error caching unlock information: %v", err)
+	}
+
+	return nil
+}
+
+// ChannelWrapper for lock search to more easily return async error data via Wait()
+// See NewPointerChannelWrapper for construction / use
+type LockChannelWrapper struct {
```

Comment:
Sounds good, I'll keep my 👀  peeled. My motivation--one I should have mentioned earlier--was that I'd like to provide the flexibility for people to handle these errors on a case-by-case basis, halting the goroutine/worker/etc if an error is deemed "fatal", and ignoring it to move on otherwise.